core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -64,126 +85,140 @@ private[fs2] object Algebra {
+            F.pure((scope, Some((output.values, f(Right(()))))))
+
+          case run: Algebra.Run[F, O, r] =>
+            val (h, t) =
+            run.values.force.splitAt(chunkSize, Some(maxSteps)) match {
+              case Left((r, chunks, _)) => (chunks, f(Right(r)))
+              case Right((chunks, tail)) => (chunks, segment(tail).transformWith(f))
+            }
+            F.pure((scope, Some((Segment.catenatedChunks(h), t))))
+
+          case uncons: Algebra.Uncons[F, x, O] =>
+            F.flatMap(F.attempt(compileUncons(scope, uncons.s, uncons.chunkSize, uncons.maxSteps))) {
+              case Right((scope, u)) => compileUncons(scope, FreeC.suspend(f(Right(u))), chunkSize, maxSteps)
+              case Left(err) => compileUncons(scope, FreeC.suspend(f(Left(err))), chunkSize, maxSteps)
+            }
+
+          case alg: AlgebraShared[F, O, _] =>
+            F.flatMap(compileShared(scope, alg)) { case (scope, r) =>
+              compileUncons(scope, f(r), chunkSize, maxSteps)
+            }
+        }
+
+      case e => sys.error("FreeC.ViewL structure must be Pure(a), Fail(e), or Bind(Eval(fx),k), was (unconcs): " + e)
+    }
+  }
+
+  private def compileFoldLoop[F[_],O,B](
     scope: CompileScope[F]
     , acc: B
     , g: (B, O) => B
-    , v: FreeC.ViewL[Algebra[F,O,?], Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]]
+    , v: FreeC[Algebra[F, O, ?], Unit]
   )(implicit F: Sync[F]): F[B] = {
+    F.flatMap(F.delay { v.viewL.get }) {
+      case done: FreeC.Pure[Algebra[F,O,?], Unit] =>
+        F.pure(acc)
+
+      case failed: FreeC.Fail[Algebra[F,O,?], Unit] =>
+        F.raiseError(failed.error)
+
+      case bound: FreeC.Bind[Algebra[F,O,?], _, Unit] =>
+        val f = bound.f.asInstanceOf[Either[Throwable,Any] => FreeC[Algebra[F,O,?], Unit]]
+        val fx = bound.fx.asInstanceOf[FreeC.Eval[Algebra[F,O,?],_]].fr
+        F.flatMap(scope.shallInterrupt) {
+          case None =>
+            fx match {
+              case output: Algebra.Output[F, O] =>
+                try {
+                  compileFoldLoop(scope, output.values.fold(acc)(g).force.run._2, g, f(Right(())))
+                }
+                catch {
+                  case err: Throwable => compileFoldLoop(scope, acc, g, f(Left(err)))
+                }
 
-    v.get match {
-      case done: FreeC.Pure[Algebra[F,O,?], Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]] => done.r match {
-        case None => F.pure(acc)
-        case Some((hd, tl)) =>
-          F.suspend {
-            try compileLoop[F,O,B](scope, hd.fold(acc)(g).force.run._2, g, uncons(tl).viewL)
-            catch { case NonFatal(e) => compileLoop[F,O,B](scope, acc, g, uncons(tl.asHandler(e)).viewL) }
-          }
-      }
-
-      case failed: FreeC.Fail[Algebra[F,O,?], _] => F.raiseError(failed.error)
+              case run: Algebra.Run[F, O, r] =>
+                try {
+                  val (r, b) = run.values.fold(acc)(g).force.run
```

Comment:
Seems like we need to use `chunkSize` and `maxSteps` here to limit this call to `run`?

Reply:
run shall always terminate after the fixes should it? Except when you define one that never terminates and as pure. If that would be the case, then, perhaps I will revert to code that stepped through uncons, but I don't think so it is necessary.

Reply:
`Stream.segment(Segment.constant(0)).evalMap(Sink.showLinesStdOut).compile.drain` <-- I'd expect this to print stuff forever but as defined now, I don't think it will print anything.

Reply:
The above example works fine with this branch (because `to` uses `evalMap` which uses `flatMap` which calls `uncons`. The result is the full infinite segment, which is then lazily folded an element at a time.

I think you can ignore the need for `chunkSize` and `maxSteps` here until/if I come up with an example that fails.

Reply:
I think this shall work always, hence the uncons is implemented with `splitAt`, honoring the chunkSize and maxSteps, however the runFold is implemented in `fold`. 

Also on the ChunkSize, wouldn't that mean here that chunk is always like 256 items? Just thinking loudly what that does with Stream[F, Byte]....